symbolic-unwind/src/evaluator/mod.rs

```diff
@@ -451,6 +452,130 @@ pub enum Identifier {
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct Rule<A: RegisterValue>(Identifier, Expr<A>);
 
+/// A half-open range `[start, end)` with some `contents` attached.
+#[derive(Clone, Debug)]
+pub struct Range<A, E> {
+    /// The left endpoint of the range.
+    pub start: A,
+
+    /// The right endpoint of teh range.
+    pub end: A,
+
+    /// The contents associated with the range.
+    pub contents: E,
+}
+
+impl<A: Eq, E> PartialEq for Range<A, E> {
+    fn eq(&self, other: &Self) -> bool {
+        self.end == other.end && self.start == other.start
+    }
+}
+
+impl<A: Eq, E> Eq for Range<A, E> {}
+
+impl<A: Ord, E> core::cmp::Ord for Range<A, E> {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.end.cmp(&other.end).then(self.start.cmp(&other.start))
+    }
+}
+
+impl<A: Ord, E> PartialOrd for Range<A, E> {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// A structure containing a set of disjoint half-open ranges `[a, b)` with attached contents.
+#[derive(Clone, Debug)]
+pub struct RangeMap<A, E> {
+    inner: Vec<Range<A, E>>,
+}
```

Comment:
IMO, `Range` is a bit confusing considering there is https://doc.rust-lang.org/std/ops/struct.Range.html
Maybe `RangeMap` should be a `Vec<(std::ops::Range<A>, E)>` internally instead?

Reply:
I think that's actually a very good idea. I created my own type so that I could implement `Ord`, but that's probably overkill since we end up using `binary_search_by_key` anyway.